beam/src/Rhyolite/Task/Beam/Worker.hs

```diff
@@ -0,0 +1,150 @@
+{-# Language FlexibleContexts #-}
+{-# Language RankNTypes #-}
+{-# Language ScopedTypeVariables #-}
+{-# Language TypeApplications #-}
+{-# Language TypeFamilies #-}
+-- | Utility functions for creating worker threads, and Beam specific task workers for running SQL.
+module Rhyolite.Task.Beam.Worker where
+
+import Control.Concurrent
+import Control.Concurrent.Async
+import Control.Concurrent.Thread.Delay
+import Control.Exception.Lifted (bracket)
+import Control.Lens ((^.))
+import Control.Monad (forM)
+import Control.Monad.Cont
+import Control.Monad.Logger (askLoggerIO, MonadLoggerIO)
+import Control.Monad.Logger.Extras (Logger(..))
+import Control.Monad.Trans.Control
+import Data.Time
+import Data.Text (Text)
+
+import Database.Beam
+import Database.Beam.Backend.SQL
+import Database.Beam.Postgres
+import Database.Beam.Postgres.Full
+import Database.Beam.Postgres.Syntax
+import Database.Beam.Query.Internal (QNested)
+import Database.Beam.Schema.Tables
+
+import Database.PostgreSQL.Serializable
+import Rhyolite.DB.Beam
+import Rhyolite.Task.Beam
+
+-- | This function retrieves the first unclaimed task from the table (passed as an argument).
+-- It then assigns this task to itself, and runs the action passed to it.
+-- Once it has the output of that action, it sets the output as the result of the task,
+-- thereby marking it as completed.
+taskWorker ::
+  forall m be db f table a b.
+  ( MonadIO m, MonadLoggerIO m, Database be db
+  , Beamable table, Table table, HasTask table a b
+  , FromBackendRow be (PrimaryKey table Identity)
+  , FieldsFulfillConstraint (HasSqlEqualityCheck be) (PrimaryKey table)
+  , FieldsFulfillConstraint (HasSqlValueSyntax PgValueSyntax) (PrimaryKey table)
+  , FromBackendRow be a, HasSqlValueSyntax PgValueSyntax b
+  , be ~ Postgres, f ~ QExpr be (QNested (QNested QBaseScope)))
+  => Connection -- ^ Connection to the database
+  -> DatabaseEntity be db (TableEntity table) -- ^ A table containing embedded tasks.
+  -> f Bool -- ^ A filter for selecting tasks from the table.
+  -> (a -> Serializable (m (Serializable b))) -- ^ The action, whose output is set as the result of the task.
+  -> Text -- ^ Name of the worker
+  -> m Bool
+taskWorker dbConn table ready go workerName = do
+  logger <- askLoggerIO
+  checkedOutValue <-
+    -- BEGIN Transaction
+
+    -- Do the following inside a transaction:
+    -- 1. Get the first task that is not currently checked out by any worker, lock this task
+    -- 2. Update this task to reflect that it has been checked out by current worker
+    -- 3. Perform the serializable task
+    withTransactionSerializableRunBeamPostgres dbConn $ do
+      primaryKeyAndInput <- runSelectReturningOne $ select $
+        -- Only lock one row (limit_ 1)
+        lockingAllTablesFor_ PgSelectLockingStrengthUpdate Nothing $ limit_ 1 $ do
```

Comment:
It's not clear what it means to use pessimistic locking and serializable transactions at the same time. Can you explain why we are doing this here?

Reply:
@madeline-os Thanks for pointing this out, I was not aware that Serializable transactions already ensure this. This will get updated in the next commit.